v02/event.go

```diff
@@ -0,0 +1,333 @@
+package v02
+
+import (
+	"encoding/json"
+	"net/url"
+	"reflect"
+	"strings"
+	"time"
+)
+
+// Event implements the the CloudEvents specification version 0.2
+// https://github.com/cloudevents/spec/blob/v0.2/spec.md
+type Event struct {
+	// SpecVersion is a mandatory property
+	// https://github.com/cloudevents/spec/blob/v0.2/spec.md#cloudeventsversion
+	SpecVersion string `json:"specversion" cloudevents:"ce-specversion,required"`
+	// EventType is a mandatory property
+	// https://github.com/cloudevents/spec/blob/v0.2/spec.md#eventtype
+	Type string `json:"type" cloudevents:"ce-type,required"`
+	// Source is a mandatory property
+	// https://github.com/cloudevents/spec/blob/v0.2/spec.md#source
+	Source url.URL `json:"source" cloudevents:"ce-source,required"`
+	// ID is a mandatory property
+	// https://github.com/cloudevents/spec/blob/v0.2/spec.md#eventid
+	ID string `json:"id" cloudevents:"ce-id,required"`
+	// Time is an optional property
+	// https://github.com/cloudevents/spec/blob/v0.2/spec.md#eventtime
+	Time *time.Time `json:"time,omitempty" cloudevents:"ce-time"`
+	// SchemaURL is an optional property
+	// https://github.com/cloudevents/spec/blob/v0.2/spec.md#schemaurl
+	SchemaURL url.URL `json:"schemaurl,omitempty" cloudevents:"ce-schemaurl"`
+	// ContentType is an optional property
+	// https://github.com/cloudevents/spec/blob/v0.2/spec.md#contenttype
+	ContentType string `json:"contenttype,omitempty" cloudevents:"content-type"`
+	// Data is an optional property
+	// https://github.com/cloudevents/spec/blob/v0.2/spec.md#data-1
+	Data interface{} `json:"data,omitempty" cloudevents:",body"`
+	// extension an internal map for extension properties not defined in the spec
+	extension map[string]interface{}
```

Comment:
extensions are handled different, e.g.:

```
CURL -X POST example/webhook.json \
-H 'traceparent:  00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01' \
-H 'tracestate: rojo=00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01,congo=lZWRzIHRoNhcm5hbCBwbGVhc3VyZS4=` \
-H 'content-type: application/cloudevents+json' \
-d '@sample-event.json'
```

for Distributed Tracing. See: https://github.com/cloudevents/spec/blob/master/extensions/distributed-tracing.md

Reply:
I'm a little confused by this. Are you saying that extensions' headers in the HTTP binding are not prefixed with 'ce-'? The marhallers written for the Event struct know to unpack this map into top level properties in either the binary or json representation, but the the binary marshaller assumes that extensions' headers will be prefixed with 'ce-' like all of the other cloud event properties.